File: Week0-1/credit/credit.c
//https://cs50.harvard.edu/x/2023/psets/1/credit/

#include <cs50.h>
#include <stdio.h>

long get_cardnum(void);
long length_card(long cardnum);

int main(void)
{
    //Define the card number we want to analyse
    long cardnum = get_cardnum();
//    printf("the num is : %li\n", cardnum);

    //Make the sum of even digits (x2) et odd digits (starting from the end)

    long digit = 1;  // the analized digit = cardnum %10 : it is a tracker
    long sum_odd = 0;
    long sum_even = 0;
    long restdigit = cardnum;
    while (restdigit > 0)
    {
        if (digit % 2 == 0) // check if the digit is even
        {
            if ((restdigit % 10) > 4) // to analyse digit that contains 2  digits (ie (>9 like 10 or 11...)when multiplited by 2
            {
                sum_even += ((2 * (restdigit % 10)) / 10 + (2 * (restdigit % 10)) % 10);    //if the digit is even and double
            }
            else
            {
                sum_even += (restdigit % 10) * 2;    //if the digit is even non double
            }
        }
        else // or
        {
            sum_odd += restdigit % 10;     //if the digit is odd
        }
        restdigit = restdigit / 10;
        digit++;
    }

// Make the total sum
    long total;
    total = sum_odd + sum_even;

//    printf ("odd : %li \n", sum_odd);
//    printf ("even : %li \n", sum_even);
//   printf ("total is : %li \n", total);


// Guess the brand
    long lengthcard = length_card(cardnum);
//    printf("the length is : %li\n", lengthcard);
    if (total % 10 != 0)
    {
        printf("INVALID\n");
    }

    else if (lengthcard == 15 && (cardnum / 10000000000000 == 34 || cardnum / 10000000000000 == 37))
    {
        printf("AMEX\n");
    }
    else if ((lengthcard == 13 || lengthcard == 16)  && (cardnum / 1000000000000 == 4 || cardnum / 1000000000000000 == 4))
    {
        printf("VISA\n");
    }
    else if ((lengthcard == 16)  && (cardnum / 100000000000000 >= 51 && cardnum / 100000000000000 <= 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
// FONCTIONS -----------------------------------------------------------------------------------------------------------------------------------------------------------------

//---------- Prompt the card number
long get_cardnum(void)
{
    long cardnum;
    do
    {
        cardnum = get_long("Number: ");
    }
    while (cardnum < 0);
    return cardnum;
}
//---------- Get the length of the card
long length_card(long cardnum)
{
    long lengthcard = 0;
    while (cardnum > 0)
    {
        cardnum /= 10;
        lengthcard++;
    }
    return lengthcard;
}

File: Week6/sentimental-credit/credit.py
# TODO

from cs50 import get_string

# Ask the card number
while True:
    try:
        num_card = get_string("Number: ")
        if int(num_card) > 0:
            break
    except ValueError:
        pass

# starting from the end of the card number,  multiply each even numbers by 2 and make the sum = sum 1
# make the sum of every odd numbers and make sum 2
# the total sum = sum1 + sum2 . If its last digit is 0 -> valid
# to make it easier compared with C, we can converte the num card wich is a string into a list of integer,
# sorted by reverse
num_card = num_card[::-1]

sum_even = 0
sum_odd = 0

# we can now easily make the sum of every even digits
# if the digit has 2 digits ex: 23. The result is i %10 = 3  + i // 10 = 2     total = 3 + 2 = 5
for i in num_card[1::2]:
    i = int(i) * 2
    if i >= 10:
        sum_even += i % 10 + i // 10
    else:
        sum_even += i

# we can now easily make the sum of every odd digits
for j in num_card[::2]:
    sum_odd += int(j)

total = sum_even + sum_odd

num_cardb = num_card[::-1]
print(num_cardb[0:2])

L = int(len(num_card))

if total % 10 != 0:
    print("INVALID")
elif L == 15 and ((num_cardb[0:2] == "34") or num_cardb[0:2] == "37"):
    print("AMEX")
elif L == 13 or L == 16 and (num_cardb[0:1] == "4"):
    print("VISA")
elif L == 16 and (int(num_cardb[0:2]) >= 51 and int(num_cardb[0:2]) <= 55):
    print("MASTERCARD")
else:
    print("INVALID")